d3ploy/NOInst.py

```diff
@@ -0,0 +1,265 @@
+        in supply and demand and makes the the decision to deploy facilities or not.     
+        """
+        time = self.context.time
+        diff, supply, demand = self.calc_diff(time)
+        if  diff < 0:
+            proto = random.choice(self.prototypes)
+            #proto_v = self.fac_supply[proto]
+            prod_rate = self.commodity_supply[time] / self.deployed
+            number = np.ceil(-1*diff/prod_rate)
+            self.deployed += number
+            i = 0
+            while i < number:
+                self.context.schedule_build(self, proto)
+                i+=1
+        if self.record:
+            f = open(self.demand_commod+".txt", 'a')
+            f.write("Time " + str(time) + " Deployed " + str(self.deployed) + " supply " + str(self.commodity_supply[time]) + " demand " +str(self.commodity_demand[time]) + "\n")   
+            f.close()  
+
+    def calc_diff(self, time):
+        """
+        This function calculates the different in supply and demand for a given facility
+        Parameters
+        ----------        
+        time : int
+            This is the time step that the difference is being calculated for.
+        Returns
+        -------
+        diff : double
+            This is the difference between supply and demand at [time]
+        supply : double
+            The calculated supply of the supply commodity at [time].
+        demand : double
+            The calculated demand of the demand commodity at [time]
+        """
+        try:
+            supply = self.calc_methods[self.calc_method](self, self.commodity_supply)
+        except:
+            supply = self.calc_methods['ma'](self, self.commodity_supply)
+        if not self.commodity_demand:
+            self.commodity_demand[time] = self.initial_demand
+        if self.demand_commod == 'power':
+            demand = self.demand_calc(time+1)
+            self.commodity_demand[time] = demand
+        else:
+            try:
+                demand = self.calc_methods[self.calc_method](self, self.commodity_demand)
+            except:
+                demand = self.calc_methods['ma'](self, self.commodity_demand)
+        diff = supply - demand
+        return diff, supply, demand
+
+    def extract_supply(self, agent, time, value):
+        """
+        Gather information on the available supply of a commodity over the
+        lifetime of the simulation. 
+
+        Parameters
+        ----------
+        agent : cyclus agent
+            This is the agent that is making the call to the listener.
+        time : int
+            Timestep that the call is made.
+        value : object
+            This is the value of the object being recorded in the time
+            series.
+        """
+        if time in self.commodity_supply:
+            self.commodity_supply[time] += value
+        else:
+            self.commodity_supply[time] = value
```

Comment:
It seems like commodity could be a defaultdict in order to simplify the logic here.